pkg/controller/networkpolicy/clustergroup.go

```diff
@@ -176,27 +173,62 @@ func (n *NetworkPolicyController) syncInternalGroup(key string) error {
 		return nil
 	}
-
 	grp := grpObj.(*antreatypes.Group)
-	// Find all Pods matching its selectors and update store.
-	groupSelector := grp.Selector
-	pods, _ := n.processSelector(groupSelector)
-	memberSet := controlplane.GroupMemberSet{}
-	for _, pod := range pods {
-		if pod.Status.PodIP == "" {
-			// No need to insert Pod IPAddress when it is unset.
-			continue
+	if grp.IPBlock == nil {
+		// Find all Pods matching its selectors and update store.
+		groupSelector := grp.Selector
+		pods, _ := n.processSelector(groupSelector)
+		memberSet := controlplane.GroupMemberSet{}
+		for _, pod := range pods {
+			if pod.Status.PodIP == "" {
+				// No need to insert Pod IPAddress when it is unset.
+				continue
+			}
+			memberSet.Insert(podToGroupMember(pod, true))
 		}
-		memberSet.Insert(podToGroupMember(pod, true))
+		// Update the internal Group object in the store with the Pods as GroupMembers.
+		updatedGrp := &antreatypes.Group{
+			UID:          grp.UID,
+			Name:         grp.Name,
+			Selector:     grp.Selector,
+			GroupMembers: memberSet,
+		}
+		klog.V(2).Infof("Updating existing internal Group %s with %d GroupMembers", key, len(memberSet))
+		n.internalGroupStore.Update(updatedGrp)
+	}
+	// Retrieve the ClusterGroup corresponding to this key.
+	cg, err := n.cgLister.Get(grp.Name)
+	if err != nil {
+		klog.Infof("Didn't find the ClusterGroup %s, skip processing of internal group", grp.Name)
+		return nil
+	}
+	// If a ClusterGroup is added/updated, it might have a reference in ClusterNetworkPolicy.
+	cnps, err := n.cnpInformer.Informer().GetIndexer().ByIndex(ClusterGroupIndex, cg.Name)
+	if err != nil {
+		klog.Errorf("Error retrieve ClusterNetworkPolicies corresponding to ClusterGroup %s", cg.Name)
+		return err
 	}
-	// Update the internal Group object in the store with the Pods as GroupMembers.
-	updatedGrp := &antreatypes.Group{
-		UID:          grp.UID,
-		Name:         grp.Name,
-		Selector:     grp.Selector,
-		GroupMembers: memberSet,
+	for _, obj := range cnps {
+		cnp := obj.(*secv1alpha1.ClusterNetworkPolicy)
+		// Re-process ClusterNetworkPolicies which may be affected due to updates to CG.
+		curInternalNP := n.processClusterNetworkPolicy(cnp)
+		klog.V(2).Infof("Updating existing internal NetworkPolicy %s for %s", curInternalNP.Name, curInternalNP.SourceRef.ToString())
+		key := internalNetworkPolicyKeyFunc(cnp)
+		// Lock access to internal NetworkPolicy store such that concurrent access
+		// to an internal NetworkPolicy is not allowed. This will avoid the
+		// case in which an Update to an internal NetworkPolicy object may
+		// cause the SpanMeta member to be overridden with stale SpanMeta members
+		// from an older internal NetworkPolicy.
+		n.internalNetworkPolicyMutex.Lock()
+		oldInternalNPObj, _, _ := n.internalNetworkPolicyStore.Get(key)
+		oldInternalNP := oldInternalNPObj.(*antreatypes.NetworkPolicy)
+		// Must preserve old internal NetworkPolicy Span.
+		curInternalNP.SpanMeta = oldInternalNP.SpanMeta
```

Comment:
Why this extra step? Wouldn't the SpanMeta be updated in `syncInternalNetworkPolicy` anyways when the internalNP key is popped from the queue?

Reply:
to ensure that the internal network policy store update does not generate events with no span info and preserve older data

Reply:
I see, so if store sees an internal NP update with no span, it will think that the NP does not apply to those Nodes anymore, even it's a very transient state. Nevertheless, could we wrap L205-L231 in a separate function? This syncInternalGroup is going to get a lot messier once we need to process serviceSelectors too :)

Reply:
done

Reply:
thanks